File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        if len(p)>len(s):
            return []
        res=[]
        p_char={}
        for ch in p:
            if ch not in p_char:
                p_char[ch]=1
            else:
                p_char[ch]+=1
        window_char={}
        start=0
        for end_ix,char in enumerate(s):
            if char not in p_char:
                start=end_ix+1
                window_char.clear()
            else:
                if char in window_char:
                    window_char[char]+=1

                    while window_char[char]>p_char[char]:
                        window_char[s[start]]-=1
                        start+=1
                else:
                    window_char[char]=1
                if window_char==p_char:
                    res.append(start)
        return res
                

    
        

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int k=p.size();
        int n=s.size();
        if(k>n) return {};
        vector<int> frq1(26,0),frq2(26,0),ans;
        for(char c:p)frq1[c-'a']++;
        for(int i=0;i<k;i++){
            frq2[s[i]-'a']++;
        }
        if(frq1==frq2) ans.push_back(0);
        for(int i=k;i<n;i++){
            frq2[s[i]-'a']++;
            frq2[s[i-k]-'a']--;
            if(frq2==frq1) ans.push_back(i-k+1);
        }
        return ans;
    }
};